### PR TITLE
Scriptomation yaml editor

### PR DESCRIPTION
**Pull request by thomasloven**
## Proposed change

This replaces #6986 

image

Also fixes a translation error in the script editor menu.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by thomasloven** [comment]
The code formatter really did a number on the diff for this one, but with whitespace changes hidden it looks all right.

File: src/panels/config/automation/ha-automation-editor.ts [hunk 1]
```diff
@@ -512,6 +602,27 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
     triggerAutomation(this.hass, (ev.target as any).stateObj.entity_id);
   }
 
+  private _preprocessYaml() {
+    const cleanConfig = deepClone(this._config);
+    delete cleanConfig.id;
+    return cleanConfig;
+  }
```
Comment (thomasloven) [inline src/panels/config/automation/ha-automation-editor.ts:609]: For explanation of why this is a thing: We do not want the user to mess with the `id` property, and we don't want it to be copied and pasted either.

I plan more uses for this function.

Comment (thomasloven) [reply src/panels/config/automation/ha-automation-editor.ts:609]: Actually, come to think of it those uses won't apply to scripts anyway. But maybe it should be left in there for consistency?

File: src/panels/config/automation/ha-automation-editor.ts [hunk 2]
```diff
@@ -113,6 +117,37 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
+            )}
+            graphic="icon"
+          >
+            ${this.hass.localize("ui.panel.config.automation.editor.edit_ui")}
+            ${this._mode === "gui"
+              ? html`<ha-svg-icon
+                  slot="graphic"
+                  .path=${mdiCheck}
+                ></ha-svg-icon> `
+              : ``}
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:132]: use `<ha-button-menu activatable>` and `<mwc-list-item selected activated>...</mwc-list-item>`

Comment (bramkragten) [reply src/panels/config/automation/ha-automation-editor.ts:132]: Could be besides the check

File: src/panels/config/automation/ha-automation-editor.ts [hunk 3]
```diff
@@ -512,6 +602,27 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
     triggerAutomation(this.hass, (ev.target as any).stateObj.entity_id);
   }
 
+  private _preprocessYaml() {
+    const cleanConfig = deepClone(this._config);
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:606]: Do we have to deepclone it? A shallow copy will do for just removing the id?

File: src/panels/config/script/ha-script-editor.ts [hunk 4]
```diff
@@ -419,6 +501,25 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
     this._dirty = true;
   }
 
+  private _preprocessYaml() {
+    return this._config;
+  }
+
+  private async _copyYaml() {
+    const yaml = await import("js-yaml");
+    navigator.clipboard.writeText(yaml.safeDump(this._preprocessYaml()));
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:510]: Can't we get the YAML content from the `ha-yaml-editor` component?

Comment (thomasloven) [reply src/panels/config/script/ha-script-editor.ts:510]: Well... it's a private property... though that can be changed...

`js-yaml` is most likely loaded anyway already, so it's a cheap operation.

File: src/panels/config/automation/ha-automation-editor.ts (hunk 1 above)
Comment (zsarnett) [reply src/panels/config/automation/ha-automation-editor.ts:609]: Yes I need this for the grid layout as well. I just didn't know the best way to do it. Thanks! :)

File: src/panels/config/script/ha-script-editor.ts (hunk 4 above)
Comment (bramkragten) [reply src/panels/config/script/ha-script-editor.ts:510]: Add a getter to `ha-yaml-editor` (or make it public if it is already there)

File: src/panels/config/automation/ha-automation-editor.ts (hunk 2 above)
Comment (thomasloven) [reply src/panels/config/automation/ha-automation-editor.ts:138]: Looks great!

File: src/panels/config/automation/ha-automation-editor.ts [hunk 5]
```diff
@@ -89,6 +93,10 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
 
   @internalProperty() private _entityId?: string;
 
+  @internalProperty() private _mode = "gui";
```
Comment (zsarnett) [inline src/panels/config/automation/ha-automation-editor.ts:96]:
```suggestion
  @internalProperty() private _mode: "gui" | "yaml" = "gui";
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 6]
```diff
@@ -113,6 +122,39 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
+            )}
+            graphic="icon"
+            ?activated=${this._mode === "gui"}
+          >
+            ${this.hass.localize("ui.panel.config.automation.editor.edit_ui")}
+            ${this._mode === "gui"
+              ? html`<ha-svg-icon
+                  slot="graphic"
+                  .path=${mdiCheck}
+                ></ha-svg-icon> `
```
Comment (zsarnett) [inline src/panels/config/automation/ha-automation-editor.ts:137]:
```suggestion
                ></ha-svg-icon>`
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 7]
```diff
@@ -113,6 +122,39 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
+            )}
+            graphic="icon"
+            ?activated=${this._mode === "yaml"}
+          >
+            ${this.hass.localize("ui.panel.config.automation.editor.edit_yaml")}
+            ${this._mode === "yaml"
+              ? html`<ha-svg-icon
+                  slot="graphic"
+                  .path=${mdiCheck}
+                ></ha-svg-icon> `
```
Comment (zsarnett) [inline src/panels/config/automation/ha-automation-editor.ts:152]:
```suggestion
                ></ha-svg-icon>`
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 8]
```diff
@@ -152,215 +194,270 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
+                                        `ui.panel.config.automation.editor.modes.${mode}`
+                                      ) || mode}
+                                    </paper-item>
+                                  `
+                                )}
+                              </paper-listbox>
+                            </paper-dropdown-menu-light>
+                            ${this._config.mode &&
+                            MODES_MAX.includes(this._config.mode)
+                              ? html` <paper-input
```
Comment (zsarnett) [inline src/panels/config/automation/ha-automation-editor.ts:275]:
```suggestion
                              ? html`<paper-input
```

File: src/panels/config/script/ha-script-editor.ts [hunk 9]
```diff
@@ -68,6 +71,10 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
 
   @internalProperty() private _errors?: string;
 
+  @internalProperty() private _mode = "gui";
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:74]:
```suggestion
  @internalProperty() private _mode: "gui" | "yaml" = "gui";
```

File: src/panels/config/script/ha-script-editor.ts [hunk 10]
```diff
@@ -89,17 +97,48 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+            )}
+            graphic="icon"
+            ?activated=${this._mode === "gui"}
+          >
+            ${this.hass.localize("ui.panel.config.automation.editor.edit_ui")}
+            ${this._mode === "gui"
+              ? html`<ha-svg-icon
+                  slot="graphic"
+                  .path=${mdiCheck}
+                ></ha-svg-icon> `
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:112]:
```suggestion
                ></ha-svg-icon>`
```

File: src/panels/config/script/ha-script-editor.ts [hunk 11]
```diff
@@ -89,17 +97,48 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+            )}
+            graphic="icon"
+            ?activated=${this._mode === "yaml"}
+          >
+            ${this.hass.localize("ui.panel.config.automation.editor.edit_yaml")}
+            ${this._mode === "yaml"
+              ? html`<ha-svg-icon
+                  slot="graphic"
+                  .path=${mdiCheck}
+                ></ha-svg-icon> `
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:127]:
```suggestion
                ></ha-svg-icon>`
```

File: src/panels/config/script/ha-script-editor.ts [hunk 12]
```diff
@@ -110,164 +149,215 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
-                                "ui.panel.config.script.editor.id"
-                              )}
-                              .errorMessage=${this.hass.localize(
-                                "ui.panel.config.script.editor.id_already_exists"
+          ${this._mode === "gui"
+            ? html`
+                <div
+                  class="${classMap({
+                    rtl: computeRTL(this.hass),
+                  })}"
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:157]:
```suggestion
                  class=${classMap({
                    rtl: computeRTL(this.hass),
                  })}
```

File: src/panels/config/script/ha-script-editor.ts [hunk 13]
```diff
@@ -110,164 +149,215 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+                                .label=${this.hass.localize(
+                                  "ui.panel.config.script.editor.icon"
+                                )}
+                                .name=${"icon"}
+                                .value=${this._config.icon}
+                                @value-changed=${this._valueChanged}
+                              >
+                              </ha-icon-input>
+                              ${!this.scriptEntityId
+                                ? html` <paper-input
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:194]:
```suggestion
                                ? html`<paper-input
```

File: src/panels/config/script/ha-script-editor.ts [hunk 14]
```diff
@@ -110,164 +149,215 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+                                          `ui.panel.config.script.editor.modes.${mode}`
+                                        ) || mode}
+                                      </paper-item>
+                                    `
+                                  )}
+                                </paper-listbox>
+                              </paper-dropdown-menu-light>
+                              ${this._config.mode &&
+                              MODES_MAX.includes(this._config.mode)
+                                ? html` <paper-input
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:250]:
```suggestion
                                ? html`<paper-input
```

File: src/panels/config/script/ha-script-editor.ts [hunk 15]
```diff
@@ -110,164 +149,215 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+                        </ha-config-section>
+                      `
+                    : ""}
+                </div>
+              `
+            : this._mode === "yaml"
+            ? html`
+                <ha-config-section .isWide=${false}>
+                  ${!this.narrow
+                    ? html` <span slot="header">${this._config?.alias}</span> `
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:324]:
```suggestion
                    ? html`<span slot="header">${this._config?.alias}</span>`
```

File: src/panels/config/script/ha-script-editor.ts [hunk 16]
```diff
@@ -110,164 +149,215 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+                              )}"
+                              ?disabled=${this._dirty}
+                            >
+                              ${this.hass.localize("ui.card.script.execute")}
+                            </mwc-button>
+                          </div>
+                        `
+                      : ``}
+                  </ha-card>
+                  <ha-config-section> </ha-config-section
```
Comment (zsarnett) [inline src/panels/config/script/ha-script-editor.ts:357]: ?
```suggestion
```

Comment (zsarnett) [reply src/panels/config/script/ha-script-editor.ts:357]: Actually I think part of this line is on the next line so my suggestion is bad but I don't think this is needed

Comment (thomasloven) [reply src/panels/config/script/ha-script-editor.ts:357]: I was sure I had removed this already...

File: src/panels/config/script/ha-script-editor.ts (hunk 12 above)
Comment (thomasloven) [reply src/panels/config/script/ha-script-editor.ts:157]: Well... those aren't even my fault. :)

**Pull request merged**